src/effects.ts

```diff
@@ -12,16 +12,82 @@ export function action<Action>(action: Action): Effect<Action> {
   return [dispatch => dispatch(action)];
 }
 
-export function fromPromise<Action>(
-  promise: () => Promise<Action>
+export function fromPromise<Action, Value = unknown, Err = Error>(
+  promise: () => Promise<Value>,
+  ofSuccess: (value: Value) => Action,
+  ofError: (error: Err) => Action
+): Effect<Action>
+
+export function fromPromise<Action, Value = unknown, Err = Error>(
+  promise: () => Promise<Value>,
+  ofError: (error: Err) => Action
+): Effect<Action>
+
+export function fromPromise<Action, Value, Err>(
+  promise: () => Promise<Value>,
+  ofSuccess: ((value: Value) => Action) | ((error: Err) => Action),
+  ofError?: (error: Err) => Action
 ): Effect<Action> {
-  return [dispatch => promise().then(dispatch)];
+  const handleError = ofError === undefined
```

Comment:
If I understand this function correctly, it is taking in a promise, and depending on whether `ofError` is defined conditionally uses `ofSuccess` as an error handler, or to map the value from `Value` to `Action`

It took me a little while to understand that intent, and that is largely because I feel the naming could be more clear here. But even if the arguments are named better, I believe that exposing this function signature to end users would result in a fair amount of confusion. The overloads are much clearer.

Reply:
This function is not exposed to the end users, see - https://www.typescriptlang.org/docs/handbook/functions.html#overloads

Reply:
also generated typings - https://github.com/iyegoroff/react-use-elmish/blob/prebuilt/dist/effects.d.ts

Reply:
Ah I see. I actually haven't used overloads before as I personally prefer static function definitions, but that makes sense.